### PR TITLE
Add the StashStatusPush handler

### PR DESCRIPTION
**Pull request by ahammond**
This adds a StashStatusPush handler that works with Atlassian Stash to update commits with build status.

### CODE REVIEW

**Comment by sa2ajj** [comment]
pyflakes error:

```
master/buildbot/status/status_stash.py:9: 'Property' imported but unused
```

File: master/docs/manual/cfg-statustargets.rst [hunk 1]
```diff
@@ -1561,6 +1561,37 @@ In case any of `repoOwner`, `repoName` or `sha` returns `None`, `False` or empty
 
 You can define custom start and end build messages using the `startDescription` and `endDescription` optional interpolation arguments.
 
+StashStatusPush
+~~~~~~~~~~~~~~~
+
+.. @cindex StashStatusPush
```
Comment (sa2ajj) [inline master/docs/manual/cfg-statustargets.rst]: `@cindex` is a remnant of texinfo format.

Please see other classes for an example of `.. index::`

**Comment by sa2ajj** [comment]
There are three things I'd like to bring up:

* at least one test case
* update of release notes
* since we have plugins in eight, update of `master/setup.py` to inlcude this new status target there

File: master/buildbot/status/status_stash.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    def __init__(self, host, user, password):
+        """
+        :param host: The domain name or IP of the stash host.
+        :param user: The stash user to log in as using http basic auth.
+        :param password: The password to use with the stash user.
+        :return:
+        """
+        StatusReceiverMultiService.__init__(self)
+        self.base_url = 'https://' + host + '/rest/build-status/1.0/commits/'
```
Comment (sa2ajj) [inline master/buildbot/status/status_stash.py]: I'd rather see a format string as below.

File: master/buildbot/status/status_stash.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        self.auth = b64encode('{0}:{1}'.format(user, password))
```
Comment (sa2ajj) [inline master/buildbot/status/status_stash.py]: Master needs to support python 2.6 where `{0}` format string will not work.

File: master/buildbot/status/status_stash.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        StatusReceiverMultiService.startService(self)
+
+    def stopService(self):
+        self.summaryUnsubscribe()
+
+    def builderAdded(self, name, builder):
+        return self
+
+    def buildStarted(self, builderName, build):
+        self.send(builderName, build, 'INPROGRESS')
```
Comment (sa2ajj) [inline master/buildbot/status/status_stash.py]: I'd suggest to make these special names -- `INPROGRESS`, `SUCCESSFUL`, `FAILED` -- global private constants.

File: master/docs/manual/cfg-statustargets.rst [hunk 5]
```diff
@@ -1561,6 +1561,37 @@ In case any of `repoOwner`, `repoName` or `sha` returns `None`, `False` or empty
...
+    from buildbot.status.status_stash import StashStatusPush
+
+    ss = StashStatusPush('stash_host.example.com',
+                         'stash_username',
+                         'secret_password')
+
+    c['status'].append(ss)
+
+
+:class:`StashStatusPush` publishes build status using `Stash Build Integration REST API <https://developer.atlassian.com/static/rest/stash/3.6.0/stash-build-integration-rest.html>`_.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-statustargets.rst]: I'd somehow emphasize that the status is published for each **build**

**Comment by sa2ajj** [comment]
Thank you for your PR :)  Please let me know if you need help.

File: master/docs/manual/cfg-statustargets.rst (hunk 1 above)
Comment (ahammond) [inline master/docs/manual/cfg-statustargets.rst]: I actually don't know squat about rst, so I just followed the example for the GitHubStatus entry above.

Comment (sa2ajj) [inline master/docs/manual/cfg-statustargets.rst]: Yeah, sorry about that: the docs need to be cleaned up :/

Comment (ahammond) [inline master/docs/manual/cfg-statustargets.rst]: I'm going to leave that to be consistent with the rest of the file then.

**Comment by ahammond** [comment]
@sa2ajj I think that covers all your comments.

File: master/docs/manual/cfg-statustargets.rst [hunk 6]
```diff
@@ -1542,7 +1542,9 @@ GitHubStatus
     c['builders'].append(buildbot_bbtools)
     c['status'].append(gs)
 
-:class:`GitHubStatus` publishes a build status using `GitHub Status API <http://developer.github.com/v3/repos/statuses>`_.
+:class:`GitHubStatus` publishes build status using `GitHub Status API <http://developer.github.com/v3/repos/statuses>`_.
+The build status is published to a specific commit SHA in Stash.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-statustargets.rst]: I think it's a typo here: Stash.

**Comment by sa2ajj** [comment]
Yes, it does.  Thank you! :)

There's a typo introduced though.

And I'll wait for Travis to show something green :)

**Comment by sa2ajj** [comment]
Thanks.  Now it's just Travis.

**Comment by sa2ajj** [comment]
Problem with a test

```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/trial/runner.py", line 561, in loadPackage
    module = modinfo.load()
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/modules.py", line 383, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/_reflectpy3.py", line 268, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/twisted/python/_reflectpy3.py", line 211, in _importAndCheckStack
    reraise(excValue, excTraceback)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_status_stash.py", line 18, in <module>
    from buildbot.status.status_stash import StashStatusPush, INPROGRESS, SUCCESSFUL, FAILED
  File "/home/travis/build/buildbot/buildbot/master/buildbot/status/status_stash.py", line 14, in <module>
    from twisted.web.client import Agent, readBody
exceptions.ImportError: cannot import name readBody

buildbot.test.unit.test_status_stash
```

**Comment by sa2ajj** [comment]
It looks like it works with Twisted v12.x, while fails for Twisted 13 & 14

**Comment by ahammond** [comment]
Strange. I tested it on Twisted 13.2

**Comment by ahammond** [comment]
http://twistedmatrix.com/documents/current/api/twisted.web.client.readBody.html

**Comment by sa2ajj** [comment]
I said "it looks like" -- based on the matrix Travis shows.

**Comment by sa2ajj** [comment]
Interestingly, the test succeed for me in this virtual env:

```
argparse (1.2.1)
buildbot (0.8.11-pre-95-g2dbac6b)
decorator (3.4.0)
Jinja2 (2.7.3)
MarkupSafe (0.23)
mock (1.0.1)
pip (1.5.6)
python-dateutil (2.4.0)
setuptools (5.5.1)
six (1.9.0)
SQLAlchemy (0.7.10)
sqlalchemy-migrate (0.7.2)
Tempita (0.5.2)
Twisted (14.0.2)
wsgiref (0.1.2)
zope.interface (4.1.2)
```

File: master/buildbot/status/status_stash.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    def builderAdded(self, name, builder):
+        return self
+
+    def buildStarted(self, builderName, build):
+        self.send(builderName, build, INPROGRESS)
+        return self
+
+    def buildFinished(self, builderName, build, results):
+        self.send(builderName, build,
+                  SUCCESSFUL if results == SUCCESS else FAILED)
```
Comment (ahammond) [inline master/buildbot/status/status_stash.py]: @sa2ajj this doesn't appear to be working. I have builds that are passing that are being reported as FAILED. So... another issue. I'll look into it now.

**Comment by ahammond** [comment]
@sa2ajj ok, I merged it to catch up with eight. UTs pass regularly on the environments I've tried it in (2.6 and 2.7 with twisted 14). I tested it against a trivial repository with a build that always succeeds and it correctly marked it as successful, so I'm guessing the false negative reports are something to do with a big ugly build that I was testing it against more than anything else.

**Comment by sa2ajj** [comment]
@ahammond Please do a rebase.  Such merges make it difficult to impossible to port the change should it be relevant.

Please let me know if you need help.

**Comment by ahammond** [comment]
Rebased, and squashed to a single commit for pretty history. I also changed the init() to take a base_url rather than host so that people can decide for themselves if they prefer http or https. Updated docs to reflect this change.

**Comment by sa2ajj** [comment]
@ahammond thank you.

We still have this strange error :(  I'm trying to figure out what's wrong.

**Comment by sa2ajj** [comment]
We have pyflakes errors:

```
pyflakes master/buildbot slave/buildslave
master/buildbot/test/unit/test_status_stash.py:20: 'fakedb' imported but unused
master/buildbot/test/unit/test_status_stash.py:21: 'fakemaster' imported but unused
master/buildbot/test/unit/test_status_stash.py:24: 'defer' imported but unused
```

And @skelly said that `readBody` is only available w/ Twisted v13.1.0+

(I'm checking why builds w/ Twisted 12.0 did not fail)

**Comment by sa2ajj** [comment]
I still do not quite understand why the test is not failing in Travis (looks like it's skipped in a fancy way for some reason).

I run the test locally w/ Twisted 12.0.0 and it does indeed fail.

Since the requirement is Twisted >= 11.0.0, this somehow needs to be accomodated (@djmitche, do we have a reason to be still compatible w/ it?)

**Comment by ahammond** [comment]
I removed the unused imports from the test code.

**Comment by sa2ajj** [comment]
So I made an experiment (GH-1529) and it turned out that installing `txgithub` when Twisted is too old for it (the requirement is '>= 12.3.0), the latter gets upgraded to the latest one and the test passes.

@ahammond the code need to be adapted so it would work with versions of Twisted supported by Buildbot.

**Comment by sa2ajj** [comment]
@ahammond alternative is to check gracefully if the function is available and refuse to work if it does not.

**Comment by djmitche** [comment]
I think this is a relatively straightforward fix.  The problem with requiring a newer Twisted is that it causes problems for downstream packager where they must wait for those newer Twisted versions to stabilize.  If this turns out to be a difficult fix, we can go down that route, but in general I try to hold out as long as practical before increasing python requirements.

**Comment by ahammond** [comment]
I wrapped the readBody() import in a try with a stub function that, in the event of an Stash API failure, will tell the user to upgrade to Twisted>=13.1 if they want verbose error messages. This should solve the import issue.

**Comment by sa2ajj** [comment]
Thank you, @ahammond 

There's a little problem with `pep8` being too new.  So I created GH-1537 where the corresponding change from master is cherry-picked.

File: master/buildbot/status/status_stash.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        (sha, ) = yield defer.gatherResults([build.render(self._sha), ])
+        build_url = build.builder.status.getURLForThing(build)
+        body = dumps({'state': status, 'key': builderName, 'url': build_url})
+        stash_uri = self.base_url + sha
+        agent = Agent(reactor)
+        d = agent.request(
+            method='POST',
+            uri=bytes(stash_uri),
+            headers=Headers({
+                'Content-Type': ['application/json; charset=utf-8',],
```
Comment (sa2ajj) [inline master/buildbot/status/status_stash.py]: One missing space:

```
pep8 --config=common/pep8rc master slave
master/buildbot/status/status_stash.py:101:67: E231 missing whitespace after ','
                'Content-Type': ['application/json; charset=utf-8',],
                                                                  ^
1       E231 missing whitespace after ','
1
make: *** [pep8] Error 1
```

Do you need a comma here at all?

**Comment by ahammond** [comment]
added pep8 silencing space.

**Comment by sa2ajj** [comment]
Thank you very much for being patient and addressing review comments.

And thank you for your contribution, @ahammond

**Pull request merged**